ops/pebble.py

```diff
@@ -67,22 +67,34 @@ def http_open(self, req):
         return self.do_open(_UnixSocketConnection, req, socket_path=self.socket_path)
 
 
-# Matches yyyy-mm-ddTHH:MM:SS.sss[-+]zz(:)zz
+# Matches yyyy-mm-ddTHH:MM:SS.sssZZZ
 _TIMESTAMP_RE = re.compile(
-    r'(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})\.(\d+)([-+])(\d{2}):?(\d{2})')
+    r'(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})\.(\d+)(.*)')
+
+# Matches [-+]HH:MM or [-+]HHMM
+_TIMEOFFSET_RE = re.compile(r'([-+])(\d{2}):?(\d{2})')
 
 
 def _parse_timestamp(s):
     """Parse timestamp from Go-encoded JSON (which uses 9 decimal places for seconds)."""
     match = _TIMESTAMP_RE.match(s)
```

Comment:
I wonder if this is something where we'd rather first try to just datetime.datetime.strptime() and only do the rest of this if that fails, rather than doing the regex test first.

The nice thing is that if new versions of python start supporting more of ISO 8601, we don't have to keep second guessing them.

Given that Python 3.7 introduces:
https://docs.python.org/3/library/datetime.html#datetime.datetime.fromisoformat

Is it worth checking to see if that method is available, and using it instead of strptime? (note that it doesn't looks like isoformat supports nanoseconds either, though.)


```
Python 3.8.5 (default, Jul 28 2020, 12:59:40)
>>> import datetime
>>> datetime.datetime.fromisoformat('2000-01-02T03:04:05.006000+0400')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Invalid isoformat string: '2000-01-02T03:04:05.006000+0400'
>>> datetime.datetime.fromisoformat('2000-01-02T03:04:05.006000+04:00')
>>> datetime.datetime.fromisoformat('2000-01-02T03:04:05.006Z')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Invalid isoformat string: '2000-01-02T03:04:05.006Z'
```

Maybe not.

Reply:
Note that from:
https://en.wikipedia.org/wiki/ISO_8601

`20210210T215346Z` is also a valid iso 8601 time.

Reply:
Yeah, Go produces RFC3339 timestamps, which is a much stricter subset of ISO8601 (thankfully). [See grammar here.](https://tools.ietf.org/html/rfc3339#section-5.6)

I don't think we should use `fromisoformat` at all. It's actually pretty dumb and mis-named, IMO. It only supports the subset of ISO8601 that Python's `isoformat` method produces, hence all the limitations. It's even [documented](https://docs.python.org/3/library/datetime.html#datetime.datetime.fromisoformat) with a "caution" to this effect.

And we can't really use `strptime` easily: 1) we still have to reduce the 9 digit fractional seconds to 6 with a regex or other string parsing, and 2) we'd have to try the various formats as the `'.'` and fractional seconds are optional, and 3) `%z` only supports a literal `Z` from Python 3.7 onwards. It's all very messy.

In any case, I've refactored to use the `datetime()` constructor directly, rather than formatting and re-parsing via `strptime`. Seemed more direct and avoids any `strptime` quirks between Python versions.